### Data Wrangling using R

#### Step 1: Loading the packages to be used. N/B: I already have them installed

In [1]:
library(tidyverse)  
library(dplyr)  
library(data.table)  
library(lubridate)
library(readr)

── Attaching packages ─────────────────────────────────────────────────────────────────────────────── tidyverse 1.3.1 ──

✔ ggplot2 3.3.6     ✔ purrr   0.3.4
✔ tibble  3.1.7     ✔ dplyr   1.0.9
✔ tidyr   1.2.0     ✔ stringr 1.4.0
✔ readr   2.1.2     ✔ forcats 0.5.1

── Conflicts ────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()


Attaching package: 'data.table'


The following objects are masked from 'package:dplyr':

    between, first, last


The following object is masked from 'package:purrr':

    transpose



Attaching package: 'lubridate'


The following objects are masked from 'package:data.table':

    hour, isoweek, mday, minute, month, quarter, second, wday, week,
    yday, year


The following objects are masked from 'package:base':

    date, intersect, setdiff, union




#### Step 2: Importing and merging the CSV files to be used for my analysis
I have 12 CSV files to combine for my analysis (each file for Cyclistic's bike ride details for each month of the year). I already previewed the files in Excel and confirmed that all 12 files have the same columns.I could import them one by one, but I choose to import all 12 files at once. This is to make my code scalable (where there may be much more datasets to combine. ie. 30 or more).

To do this, I create a folder (Cyclistic_csv_files) containing all the CSV files I want to use for my analysis, and then set my working directory to the folder, and use the list.files() function to list all files in the folder with the extension ".csv".  
Then I use the lapply() function to loop over that list (containing all my files) and read each file with the read_csv() function.

In [2]:
setwd("C:\\Users\\ETIABA CHAMBER'S\\Documents\\Data_Analytics\\My_portfolio\\Cyclistic_csv_files")
myfiles <- list.files(pattern ='*.csv')
df.list <- lapply(myfiles, read_csv)

Rows: 84776 Columns: 13
── Column specification ────────────────────────────────────────────────────────────────────────────────────────────────
Delimiter: ","
chr  (5): ride_id, rideable_type, start_station_name, end_station_name, memb...
dbl  (6): start_station_id, end_station_id, start_lat, start_lng, end_lat, e...
dttm (2): started_at, ended_at

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 622361 Columns: 13
── Column specification ────────────────────────────────────────────────────────────────────────────────────────────────
Delimiter: ","
chr  (5): ride_id, rideable_type, start_station_name, end_station_name, memb...
dbl  (6): start_station_id, end_station_id, start_lat, start_lng, end_lat, e...
dttm (2): started_at, ended_at

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet t

In [3]:
#random check to see that the right CSV files were imported
head(myfiles[3])
head(myfiles[6])
head(myfiles[12])

[1] "Dec.2020-divvy-tripdata.csv"

[1] "Jul.2020-divvy-tripdata.csv"

[1] "Sep.2020-divvy-tripdata.csv"

In [4]:
#random check to view the tibbles of some of my imported datasets
read.csv(myfiles[3], nrows = 6)
read.csv(myfiles[6], nrows = 6)
read.csv(myfiles[12], nrows = 6)

ride_id          rideable_type started_at          ended_at           
1 70B6A9A437D4C30D classic_bike  2020-12-27 12:44:29 2020-12-27 12:55:06
2 158A465D4E74C54A electric_bike 2020-12-18 17:37:15 2020-12-18 17:44:19
3 5262016E0F1F2F9A electric_bike 2020-12-15 15:04:33 2020-12-15 15:11:28
4 BE119628E44F871E electric_bike 2020-12-15 15:54:18 2020-12-15 16:00:11
5 69AF78D57854E110 electric_bike 2020-12-22 12:08:17 2020-12-22 12:10:59
6 C1DECC4AB488831C electric_bike 2020-12-22 13:26:37 2020-12-22 13:34:50
  start_station_name         start_station_id end_station_name         
1 Aberdeen St & Jackson Blvd 13157            Desplaines St & Kinzie St
2                               NA                                     
3                               NA                                     
4                               NA                                     
5                               NA                                     
6                               NA                                     
  end_station_id start_lat start_lng end_lat  end_lng   member_casual
1 TA1306000003   41.87773  -87.65479 41.88872 -87.64445 member       
2                41.93000  -87.70000 41.91000 -87.70000 member       
3                41.91000  -87.69000 41.93000 -87.70000 member       
4                41.92000  -87.70000 41.91000 -87.70000 member       
5                41.80000  -87.59000 41.80000 -87.59000 member       
6                41.80000  -87.59000 41.78000 -87.60000 member

ride_id          rideable_type started_at          ended_at           
1 762198876D69004D docked_bike   2020-07-09 15:22:02 2020-07-09 15:25:52
2 BEC9C9FBA0D4CF1B docked_bike   2020-07-24 23:56:30 2020-07-25 00:20:17
3 D2FD8EA432C77EC1 docked_bike   2020-07-08 19:49:07 2020-07-08 19:56:22
4 54AE594E20B35881 docked_bike   2020-07-17 19:06:42 2020-07-17 19:27:38
5 54025FDC7440B56F docked_bike   2020-07-04 10:39:57 2020-07-04 10:45:05
6 65636B619E24257F docked_bike   2020-07-28 16:33:03 2020-07-28 16:49:10
  start_station_name            start_station_id end_station_name         
1 Ritchie Ct & Banks St         180              Wells St & Evergreen Ave 
2 Halsted St & Roscoe St        299              Broadway & Ridge Ave     
3 Lake Shore Dr & Diversey Pkwy 329              Clark St & Wellington Ave
4 LaSalle St & Illinois St      181              Clark St & Armitage Ave  
5 Lake Shore Dr & North Blvd    268              Clark St & Schiller St   
6 Fairbanks St & Superior St    635              Wells St & Concord Ln    
  end_station_id start_lat start_lng end_lat  end_lng   member_casual
1 291            41.90687  -87.62622 41.90672 -87.63483 member       
2 461            41.94367  -87.64895 41.98404 -87.66027 member       
3 156            41.93259  -87.63643 41.93650 -87.64754 casual       
4  94            41.89076  -87.63170 41.91831 -87.63628 casual       
5 301            41.91172  -87.62680 41.90799 -87.63150 member       
6 289            41.89575  -87.62010 41.91213 -87.63466 casual

ride_id          rideable_type started_at          ended_at           
1 2B22BD5F95FB2629 electric_bike 2020-09-17 14:27:11 2020-09-17 14:44:24
2 A7FB70B4AFC6CAF2 electric_bike 2020-09-17 15:07:31 2020-09-17 15:07:45
3 86057FA01BAC778E electric_bike 2020-09-17 15:09:04 2020-09-17 15:09:35
4 57F6DC9A153DB98C electric_bike 2020-09-17 18:10:46 2020-09-17 18:35:49
5 B9C4712F78C1AE68 electric_bike 2020-09-17 15:16:13 2020-09-17 15:52:55
6 378BBCE1E444EB80 electric_bike 2020-09-17 18:37:04 2020-09-17 19:23:28
  start_station_name             start_station_id end_station_name          
1 Michigan Ave & Lake St          52              Green St & Randolph St    
2 W Oakdale Ave & N Broadway      NA              W Oakdale Ave & N Broadway
3 W Oakdale Ave & N Broadway      NA              W Oakdale Ave & N Broadway
4 Ashland Ave & Belle Plaine Ave 246              Montrose Harbor           
5 Fairbanks Ct & Grand Ave        24              Fairbanks Ct & Grand Ave  
6 Clark St & Armitage Ave         94                                        
  end_station_id start_lat start_lng end_lat  end_lng   member_casual
1 112            41.88669  -87.62356 41.88357 -87.64873 casual       
2  NA            41.94000  -87.64000 41.94000 -87.64000 casual       
3  NA            41.94000  -87.64000 41.94000 -87.64000 casual       
4 249            41.95606  -87.66892 41.96398 -87.63822 casual       
5  24            41.89186  -87.62101 41.89135 -87.62032 casual       
6  NA            41.91826  -87.63636 41.88000 -87.62000 casual

Now that I have confirmed that the right CSV files were imported, its time for me to merge the data sets together into one. I have already previewed the files in Excel and confirmed that all 12 CSV files have the same number of columns and same column names. So I'll use the rbindlist() function from data.table to combine the datasets together by binding them into one dataframe, df (since all the datasets in my list have the same column number and names).

In [5]:
df <- rbindlist(df.list)

Then I use the str() function to get a summary of my new dataframe df. 

In [6]:
str(df)

Classes 'data.table' and 'data.frame':	3489748 obs. of  13 variables:
 $ ride_id           : chr  "A847FADBBC638E45" "5405B80E996FF60D" "5DD24A79A4E006F4" "2A59BBDF5CDBA725" ...
 $ rideable_type     : chr  "docked_bike" "docked_bike" "docked_bike" "docked_bike" ...
 $ started_at        : POSIXct, format: "2020-04-26 17:45:14" "2020-04-17 17:08:54" ...
 $ ended_at          : POSIXct, format: "2020-04-26 18:12:03" "2020-04-17 17:17:03" ...
 $ start_station_name: chr  "Eckhart Park" "Drake Ave & Fullerton Ave" "McClurg Ct & Erie St" "California Ave & Division St" ...
 $ start_station_id  : chr  "86" "503" "142" "216" ...
 $ end_station_name  : chr  "Lincoln Ave & Diversey Pkwy" "Kosciuszko Park" "Indiana Ave & Roosevelt Rd" "Wood St & Augusta Blvd" ...
 $ end_station_id    : chr  "152" "499" "255" "657" ...
 $ start_lat         : num  41.9 41.9 41.9 41.9 41.9 ...
 $ start_lng         : num  -87.7 -87.7 -87.6 -87.7 -87.6 ...
 $ end_lat           : num  41.9 41.9 41.9 41.9 42 ...
 $ end_lng

It has a total of 3,489,748 rows and 13 columns. 

#### Step 3: Cleaning the data
Now that I have successfully imported and combined my datasets into one dataframe, the next step is for me to clean the data.  
I will begin by taking care of any duplicate entries. Having studied a sample of the dataset in Excel, the ride_id should be unique for each entry; so I will check to confirm if there are duplicates in my dataframe.   

##### ~Dropping duplicate ride_ids

In [7]:
#returns the number of duplicate ride_ids (if any)
sum(duplicated(df$ride_id))

[1] 209

I can see that there are 209 duplicate entries for ride_id. So I will use the code below to create a new dataframe df1 with the duplicates dropped.

In [8]:
#creating df1 with only the distinct ride_id entries
df1 <- distinct(df, ride_id, .keep_all = TRUE)

I can now use the str() and head() functions to view a summary of df1.

In [9]:
str(df1)
head(df1)

Classes 'data.table' and 'data.frame':	3489539 obs. of  13 variables:
 $ ride_id           : chr  "A847FADBBC638E45" "5405B80E996FF60D" "5DD24A79A4E006F4" "2A59BBDF5CDBA725" ...
 $ rideable_type     : chr  "docked_bike" "docked_bike" "docked_bike" "docked_bike" ...
 $ started_at        : POSIXct, format: "2020-04-26 17:45:14" "2020-04-17 17:08:54" ...
 $ ended_at          : POSIXct, format: "2020-04-26 18:12:03" "2020-04-17 17:17:03" ...
 $ start_station_name: chr  "Eckhart Park" "Drake Ave & Fullerton Ave" "McClurg Ct & Erie St" "California Ave & Division St" ...
 $ start_station_id  : chr  "86" "503" "142" "216" ...
 $ end_station_name  : chr  "Lincoln Ave & Diversey Pkwy" "Kosciuszko Park" "Indiana Ave & Roosevelt Rd" "Wood St & Augusta Blvd" ...
 $ end_station_id    : chr  "152" "499" "255" "657" ...
 $ start_lat         : num  41.9 41.9 41.9 41.9 41.9 ...
 $ start_lng         : num  -87.7 -87.7 -87.6 -87.7 -87.6 ...
 $ end_lat           : num  41.9 41.9 41.9 41.9 42 ...
 $ end_lng

ride_id          rideable_type started_at          ended_at           
1 A847FADBBC638E45 docked_bike   2020-04-26 17:45:14 2020-04-26 18:12:03
2 5405B80E996FF60D docked_bike   2020-04-17 17:08:54 2020-04-17 17:17:03
3 5DD24A79A4E006F4 docked_bike   2020-04-01 17:54:13 2020-04-01 18:08:36
4 2A59BBDF5CDBA725 docked_bike   2020-04-07 12:50:19 2020-04-07 13:02:31
5 27AD306C119C6158 docked_bike   2020-04-18 10:22:59 2020-04-18 11:15:54
6 356216E875132F61 docked_bike   2020-04-30 17:55:47 2020-04-30 18:01:11
  start_station_name                  start_station_id
1 Eckhart Park                        86              
2 Drake Ave & Fullerton Ave           503             
3 McClurg Ct & Erie St                142             
4 California Ave & Division St        216             
5 Rush St & Hubbard St                125             
6 Mies van der Rohe Way & Chicago Ave 173             
  end_station_name            end_station_id start_lat start_lng end_lat
1 Lincoln Ave & Diversey Pkwy 152            41.8964   -87.6610  41.9322
2 Kosciuszko Park             499            41.9244   -87.7154  41.9306
3 Indiana Ave & Roosevelt Rd  255            41.8945   -87.6179  41.8679
4 Wood St & Augusta Blvd      657            41.9030   -87.6975  41.8992
5 Sheridan Rd & Lawrence Ave  323            41.8902   -87.6262  41.9695
6 Streeter Dr & Grand Ave     35             41.8969   -87.6217  41.8923
  end_lng  member_casual
1 -87.6586 member       
2 -87.7238 member       
3 -87.6230 member       
4 -87.6722 member       
5 -87.6547 casual       
6 -87.6120 member

In [10]:
#double-checking that df1 is free of duplicate ride_ids
sum(duplicated(df1$ride_id))

[1] 0

The above code returned 0, meaning that there are no duplicate ride_id entries in df1.

##### ~Dropping irrelevant columns
The next step I will undertake in my data cleaning process is to drop the columns that would not be useful for my analysis. Having studied a sample of the data in Excel, I can tell that the following 6 columns would not be useful in my analysis, and I would be dropping them:
  i.Start_station_id
 ii.End_station_id
iii.Start_lat
 iv.Start_lng
  v.End_lat
 vi.End_lng

In [11]:
#dropping the above mentioned columns
df2 <- select(df1, -6, -8:-12)
str(df2)

Classes 'data.table' and 'data.frame':	3489539 obs. of  7 variables:
 $ ride_id           : chr  "A847FADBBC638E45" "5405B80E996FF60D" "5DD24A79A4E006F4" "2A59BBDF5CDBA725" ...
 $ rideable_type     : chr  "docked_bike" "docked_bike" "docked_bike" "docked_bike" ...
 $ started_at        : POSIXct, format: "2020-04-26 17:45:14" "2020-04-17 17:08:54" ...
 $ ended_at          : POSIXct, format: "2020-04-26 18:12:03" "2020-04-17 17:17:03" ...
 $ start_station_name: chr  "Eckhart Park" "Drake Ave & Fullerton Ave" "McClurg Ct & Erie St" "California Ave & Division St" ...
 $ end_station_name  : chr  "Lincoln Ave & Diversey Pkwy" "Kosciuszko Park" "Indiana Ave & Roosevelt Rd" "Wood St & Augusta Blvd" ...
 $ member_casual     : chr  "member" "member" "member" "member" ...
 - attr(*, ".internal.selfref")=<externalptr> 


##### ~Taking care of missing values
Now that I have dropped the columns mentioned, my next step would be to take care of missing values in the columns of my dataframe.  
The lines of code below would check to see if there are missing values in df2, and tell me which columns (if any) contain the missing variables.

In [12]:
#returns sum of missing values in df2
sum(is.na(df2))

[1] 265416

In [13]:
#returns the names of columns where those missing values exist
names(which(colSums(is.na(df2))>0))

[1] "start_station_name" "end_station_name"

The code returned 2 columns with missing values: start_station_name and end_station_name. For the purpose of this analysis, I will fill in the missing station names with "Unknown".

In [14]:
#replace missing start_name_id and end_name_id with unknown
df2 <- df2 %>% 
  replace_na(list(start_station_name = "Unknown", end_station_name = "Unknown"))

Now that I have replaced the missing station name values with "Unknown", I will use the code below to double check that there are no more missing values in df2.

In [15]:
sum(is.na(df2))

[1] 0

The code returned 0, meaning that there are no longer missing values in df2.

##### ~Adding calculated columns
For my analysis, I want to add calculated columns such as:  
-Ride_length to calculate the length of each ride   
-Day_of_week, month, season, to check for periodic trends

In [16]:
#adding ride length
df2 <- df2 %>%
  mutate(ride_length = difftime(ended_at, started_at, units = "mins"))

#adding day of week 
df2 <- df2 %>%
  mutate(day_of_week = weekdays(started_at, abbr = TRUE))

#assigning month
df2 <- df2 %>%
  mutate(month = month(started_at, label = TRUE, abbr  = TRUE))

#appending seasons
#Ps.:I could use an ifelse() function to assign seasons to months, but I believe I can get away with a slightly shorter and
#equally straightforward code using the fct_collapse() function.

df2 <- df2 %>%
  mutate(season = fct_collapse(month, "Spring" = month.abb[3:5], 
                               "Summer" = month.abb[6:8],
                               "Autumn" = month.abb[9:11],       
                               "Winter" = month.abb[c(12,1,2)]))    

str(df2)
head(df2)

Classes 'data.table' and 'data.frame':	3489539 obs. of  11 variables:
 $ ride_id           : chr  "A847FADBBC638E45" "5405B80E996FF60D" "5DD24A79A4E006F4" "2A59BBDF5CDBA725" ...
 $ rideable_type     : chr  "docked_bike" "docked_bike" "docked_bike" "docked_bike" ...
 $ started_at        : POSIXct, format: "2020-04-26 17:45:14" "2020-04-17 17:08:54" ...
 $ ended_at          : POSIXct, format: "2020-04-26 18:12:03" "2020-04-17 17:17:03" ...
 $ start_station_name: chr  "Eckhart Park" "Drake Ave & Fullerton Ave" "McClurg Ct & Erie St" "California Ave & Division St" ...
 $ end_station_name  : chr  "Lincoln Ave & Diversey Pkwy" "Kosciuszko Park" "Indiana Ave & Roosevelt Rd" "Wood St & Augusta Blvd" ...
 $ member_casual     : chr  "member" "member" "member" "member" ...
 $ ride_length       : 'difftime' num  26.8166666666667 8.15 14.3833333333333 12.2 ...
  ..- attr(*, "units")= chr "mins"
 $ day_of_week       : chr  "Sun" "Fri" "Wed" "Tue" ...
 $ month             : Ord.factor w/ 12 levels "J

ride_id          rideable_type started_at          ended_at           
1 A847FADBBC638E45 docked_bike   2020-04-26 17:45:14 2020-04-26 18:12:03
2 5405B80E996FF60D docked_bike   2020-04-17 17:08:54 2020-04-17 17:17:03
3 5DD24A79A4E006F4 docked_bike   2020-04-01 17:54:13 2020-04-01 18:08:36
4 2A59BBDF5CDBA725 docked_bike   2020-04-07 12:50:19 2020-04-07 13:02:31
5 27AD306C119C6158 docked_bike   2020-04-18 10:22:59 2020-04-18 11:15:54
6 356216E875132F61 docked_bike   2020-04-30 17:55:47 2020-04-30 18:01:11
  start_station_name                  end_station_name            member_casual
1 Eckhart Park                        Lincoln Ave & Diversey Pkwy member       
2 Drake Ave & Fullerton Ave           Kosciuszko Park             member       
3 McClurg Ct & Erie St                Indiana Ave & Roosevelt Rd  member       
4 California Ave & Division St        Wood St & Augusta Blvd      member       
5 Rush St & Hubbard St                Sheridan Rd & Lawrence Ave  casual       
6 Mies van der Rohe Way & Chicago Ave Streeter Dr & Grand Ave     member       
  ride_length   day_of_week month season
1 26.81667 mins Sun         Apr   Spring
2  8.15000 mins Fri         Apr   Spring
3 14.38333 mins Wed         Apr   Spring
4 12.20000 mins Tue         Apr   Spring
5 52.91667 mins Sat         Apr   Spring
6  5.40000 mins Thu         Apr   Spring

##### ~Removing bad data
i).The dataframe includes a few hundred entries when bikes were taken out of docks and checked for quality by Divvy staff.In those cases, the start_station_name includes the phrase "BIKE CHECKING (LBS-WH-TEST)". I will use the lines of code below to check for those cases and drop them, since they were not actual rides.

In [17]:
#check for cases where bikes were checked for quality by Divvy staff
bike_checking<- subset(df2, grepl("BIKE CHECKING", df2$start_station_name))
head(bike_checking)

ride_id          rideable_type started_at          ended_at           
1 01FD073BD38D7552 docked_bike   2020-04-24 17:35:34 2020-04-24 17:35:39
2 CECF2063F123EC9A docked_bike   2020-04-24 17:31:57 2020-04-24 17:32:03
3 FF8783D4486B3C32 docked_bike   2020-04-24 17:41:10 2020-04-24 17:41:13
4 8DFF54760BFBEF6A docked_bike   2020-04-24 17:30:33 2020-04-24 17:30:44
5 FB8CA100BDCEE0C9 docked_bike   2020-04-24 17:59:25 2020-04-24 17:59:30
6 41E80A7BDB6409C3 docked_bike   2020-04-24 17:59:12 2020-04-24 17:59:00
  start_station_name                     end_station_name                      
1 HUBBARD ST BIKE CHECKING (LBS-WH-TEST) HUBBARD ST BIKE CHECKING (LBS-WH-TEST)
2 HUBBARD ST BIKE CHECKING (LBS-WH-TEST) HUBBARD ST BIKE CHECKING (LBS-WH-TEST)
3 HUBBARD ST BIKE CHECKING (LBS-WH-TEST) HUBBARD ST BIKE CHECKING (LBS-WH-TEST)
4 HUBBARD ST BIKE CHECKING (LBS-WH-TEST) HUBBARD ST BIKE CHECKING (LBS-WH-TEST)
5 HUBBARD ST BIKE CHECKING (LBS-WH-TEST) HUBBARD ST BIKE CHECKING (LBS-WH-TEST)
6 HUBBARD ST BIKE CHECKING (LBS-WH-TEST) HUBBARD ST BIKE CHECKING (LBS-WH-TEST)
  member_casual ride_length      day_of_week month season
1 casual         0.08333333 mins Fri         Apr   Spring
2 casual         0.10000000 mins Fri         Apr   Spring
3 casual         0.05000000 mins Fri         Apr   Spring
4 casual         0.18333333 mins Fri         Apr   Spring
5 casual         0.08333333 mins Fri         Apr   Spring
6 casual        -0.20000000 mins Fri         Apr   Spring

In [18]:
#drop entries where quality checks were made
df3 <- df2[!grepl("BIKE CHECKING", df2$start_station_name),]

ii).There are entries where the start_station_id and end_station_id are the same, and the ride length is less than one minute. In this case, the customer might have had a change of mind and docked the bike back to the station. Using the lines of code below, I will view and drop such entries, since actual rides weren't made. 

In [19]:
#check for cases where customers changed their mind
change_of_mind <- subset(df3, (df3$start_station_name == df3$end_station_name) & (df3$ride_length < 1))
head(change_of_mind)

ride_id          rideable_type started_at          ended_at           
1 EA0FBC59F73B90B6 docked_bike   2020-04-28 06:22:45 2020-04-28 06:22:48
2 8A52587DCD27DB83 docked_bike   2020-04-07 06:03:39 2020-04-07 06:03:42
3 9C9D48177CF76D59 docked_bike   2020-04-12 08:40:17 2020-04-12 08:40:54
4 29C794861AE2E12D docked_bike   2020-04-11 05:23:17 2020-04-11 05:23:23
5 DF1764DB2E9ED62B docked_bike   2020-04-23 14:45:29 2020-04-23 14:45:47
6 0DFF15F9038D6B8D docked_bike   2020-04-03 15:46:08 2020-04-03 15:46:31
  start_station_name             end_station_name               member_casual
1 Cannon Dr & Fullerton Ave      Cannon Dr & Fullerton Ave      member       
2 Wells St & Hubbard St          Wells St & Hubbard St          member       
3 Wilton Ave & Diversey Pkwy     Wilton Ave & Diversey Pkwy     member       
4 Sheffield Ave & Wellington Ave Sheffield Ave & Wellington Ave member       
5 Broadway & Cornelia Ave        Broadway & Cornelia Ave        member       
6 Racine Ave & 18th St           Racine Ave & 18th St           member       
  ride_length    day_of_week month season
1 0.0500000 mins Tue         Apr   Spring
2 0.0500000 mins Tue         Apr   Spring
3 0.6166667 mins Sun         Apr   Spring
4 0.1000000 mins Sat         Apr   Spring
5 0.3000000 mins Thu         Apr   Spring
6 0.3833333 mins Fri         Apr   Spring

In [20]:
#drop entries where customers changed their mind
df4 <- df3[!(df3$start_station_name == df3$end_station_name & df3$ride_length < 1),]

In [21]:
#checking how many entries are left 
nrow(df4)

[1] 3442601

In [22]:
df4 <- df4[!(df4$ride_length < 0),]

In [23]:
df4 <- df4[!(df4$month == "Jan" & df4$ride_length <= 5.3),]
df4 <- df4[!(df4$month == "Feb" & df4$ride_length <= 4.8),]
df4 <- df4[!(df4$month == "Mar" & df4$ride_length <= 7.4),]
df4 <- df4[!(df4$month == "Apr" & df4$ride_length <= 8.7),]
df4 <- df4[!(df4$month == "May" & df4$ride_length <= 8.4),]
df4 <- df4[!(df4$month == "Jun" & df4$ride_length <= 7.1),]

df4 <- df4[!(df4$month == "Jul" & df4$ride_length <= 10.1),]
df4 <- df4[!(df4$month == "Aug" & df4$ride_length <= 11.4),]
df4 <- df4[!(df4$month == "Sep" & df4$ride_length <= 10.0),]
df4 <- df4[!(df4$month == "Oct" & df4$ride_length <= 9.5),]
df4 <- df4[!(df4$month == "Nov" & df4$ride_length <= 4.0),]
df4 <- df4[!(df4$month == "Dec" & df4$ride_length <= 6.2),]

#### Step 4: Some initial descriptive analysis
Now that I have the columns I want to use for my analysis, I'll do some arithmetic summary on my data to view what the figures look like.
I'll look at the mean, maximum and minimum ride lengths per month in the data.  Before I do this, I will convert the ride_length column to numeric type so that I can perform numeric calculations on it. 

In [24]:
#changing column ride_length to numeric type
df4$ride_length <- as.numeric(df4$ride_length)

In [25]:
#calculating mean, maximum and minimum ride lengths per month
df4%>%
  group_by(month) %>%
    summarise(mean_ride_length = round(mean(ride_length),1),
               max_ride_length = round(max(ride_length),1), 
               min_ride_length = round(min(ride_length),1))%>%
                  
      mutate(across(where(is.numeric), ~ num(., digits = 1)))

month mean_ride_length max_ride_length                       min_ride_length
1  Jan   18.8             \0331\033\0339\033825.9  5.3           
2  Feb   28.0             \0333\033\0330\033129.2  4.8           
3  Mar   30.0             \0333\033\0331\033681.7  7.4           
4  Apr   46.0             \0335\033\0338\033720.0  8.7           
5  May   39.7             \0332\033\0338\033896.9  8.4           
6  Jun   38.3             \0334\033\0331\033271.0  7.1           
7  Jul   49.3             \0334\033\0339\033965.4 10.1           
8  Aug   42.5             \0334\033\0330\033846.4 11.4           
9  Sep   35.9             \0335\033\0334\033283.3 10.0           
10 Oct   29.8             \0333\033\0335\033724.4  9.5           
11 Nov   21.9             \0333\033\0335\033934.0  4.0           
12 Dec   20.7              \0339\033741.0                 6.2

With the table above, I have a first summary of my data by month. This brings me to the end of my data wrangling process. I could use ggplot2 to do some visualization but I have chosen to use Tableau to visualize the data and explore it further; so I will export my merged and cleaned dataset for use in Tableau. 

#### Step 5: Exporting the cleaned dataset   
To export my dataset to CSV, I use the line of code below: 

In [26]:
write.csv(df4, "C:\\Users\\ETIABA CHAMBER'S\\Documents\\Data_Analytics\\My_portfolio\\Cyclistic_cleaned_csv.csv", row.names = FALSE)

Warning message in file(file, ifelse(append, "a", "w")):
"cannot open file 'C:\Users\ETIABA CHAMBER'S\Documents\Data_Analytics\My_portfolio\Cyclistic_cleaned_csv.csv': Permission denied"


ERROR: Error in file(file, ifelse(append, "a", "w")): cannot open the connection


I have saved my dataset as a CSV file named "Cyclistic_cleaned_csv". I will then import the CSV in Tableau for further analysis and visualization.